keps/0015-dry-run.md

```diff
@@ -53,10 +53,52 @@ either as an argument for built-in admission plugins or through a dryRun
 query-parameter for dynamic webhooks, will give them a chance not to have any
 side-effect.
 
-All admission controllers will have to be verified and changed. A new flag will
-be added so that webhooks can explicitly register with dry-run support. If
-dry-run is requested on a non-supported webhook, the request will be completely
-rejected.
+All built-in admission controllers will have to be checked, and the ones with side
+effects will have to be skipped or changed to only have side effects in the non
+dry-run case. Some examples of built-in admission controllers with the possibility
+forside effects are:
+- ResourceQuota
+- EventRateLimit
+- NamespaceAutoProvision
+- (Valid|Mut)atingAdmissionWebhook
+
+To address the possibility of webhook authors relying on [side effects](https://github.com/kubernetes/website/blame/836629cb118e0f74545cc7d6d97aa6b9edfa1a16/content/en/docs/reference/access-authn-authz/admission-controllers.md#L582-L584), a new field
+will be added to ```admissionregistration.k8s.io/v1beta1.ValidatingWebhookConfiguration``` and
+```admissionregistration.k8s.io/v1beta1.MutatingWebhookConfiguration``` so that webhooks
+can explicitly register as having dry-run support.
+If dry-run is requested on a non-supported webhook, the request will be completely rejected,
+as a 400: Bad Request, and the name of the webhook/webhooks which don't support dry run will be
+listed. This field will be defaulted to true and deprecated in v1, and completely removed in v2.
+All webhooks registered with v2 will be assumed to support dry run.
+```golang
+// admissionregistration.k8s.io/v1beta1
+...
+type Webhook struct {
+	...
+	// SupportsDryRun defines whether this webhook will correctly handle dryRun requests.
+	// If false, any dryRun requests to resources/subresources this webhook applies to
+	// will be completely rejected and the webhook will not be called.
+	// Defaults to false.
+	// +optional
+	SupportsDryRun *bool `json:"supportsDryRun,omitempty" protobuf:"varint,6,number,opt,name=supportsDryRun"`
```

Comment:
`DryRunnable *bool`? Per https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#naming-conventions:

> The name of a field expressing a boolean property called 'fooable' should be called Fooable, not IsFooable.

That doc also says to think twice about bool fields. Having thought twice about it here, I think subdividing aspects of an admission plugin's dry runnability is too fine-grained, and a binary "supported" / "not supported" is better (especially since we want to default this on in future versions and require full support.

Reply:
I'll add some justification for this to the doc

Reply:
Do you think it should be DryRunnable?